mohammad/JS/JS-dataFetchingAndDOM/search.js

```diff
@@ -0,0 +1,114 @@
+function addSearchedDiv(myTitle, myBody, myAuthor) {
+  const DIV = document.createElement("div");
+  DIV.setAttribute(
+    "style",
+    `
+  border: 1px solid #555;
+  padding:2rem; margin:2rem;
+  border-radius:5rem;
+  `
+  );
+
+  const title = document.createElement("p");
+  title.textContent = myTitle;
+
+  const body = document.createElement("p");
+  body.textContent = myBody;
+
+  const author = document.createElement("p");
+  author.textContent = myAuthor;
+
+  DIV.appendChild(title);
+  DIV.appendChild(body);
+  DIV.appendChild(author);
+
+  return DIV;
+}
+
+const url = "https://jsonplaceholder.typicode.com";
+async function whoIsAuthor(userID) {
+  const result = await fetch(`${url}/users/${userID}`).then((response) => {
+    return response.json();
+  });
+  return result["name"];
+}
+
+async function searchByAuthor(query) {
+  const result = await fetch(`${url}/users?name=${query}`).then((response) => {
+    return response.json();
+  });
+  if (result["length"] != 0) {
+    const authorID = result[0]["id"];
+
+    const resultPost = await fetch(`${url}/posts?userId=${authorID}`).then(
+      (response) => {
+        return response.json();
+      }
+    );
+
+    return resultPost;
+  }
+
+  return [];
+}
+
+async function searchByTitle(query) {
+  const result = await fetch(`${url}/posts?title=${query}`).then((response) => {
+    return response.json();
+  });
+
+  return result;
+}
+
+async function searchByBody(query) {
+  const result = await fetch(`${url}/posts?body=${query}`).then((response) => {
+    return response.json();
+  });
+
+  return result;
+}
+
+async function apiSearch(query) {
+  const responseDiv = document.getElementById("response");
+  responseDiv.innerHTML = "";
+
+  Promise.all([
+    searchByAuthor(query),
+    searchByTitle(query),
+    searchByBody(query),
+  ]).then((result) => {
+    result.forEach((item) => {
+      if (item["length"] != 0) {
+        item.forEach(async function (element) {
+          const authorName = await whoIsAuthor(element["userId"]);
+          responseDiv.appendChild(
+            addSearchedDiv(element["title"], element["body"], authorName)
+          );
+        });
+      }
+    });
+  });
+}
+
+//debounce function is something like a loop that call a function in a specific period
+function debounce(func, delay) {
```

Comment:
Debouncing wasn't part of your issue and is a sign of being a copycat. I don't care about it btw, it's totally up to you.

Reply:
In the search section, we need debouncing to call the apiSearch function that the user may enter a different input every few seconds.